### PR TITLE
Add seccomp GA version skew for pods

### PR DESCRIPTION
**Pull request by saschagrunert**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
    
This adds a new conversion function to the pod strategy to handle the
seccomp version skew strategy:
    
https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190717-seccomp-ga.md#version-skew-strategy
   
**Which issue(s) this PR fixes**:

None

**Special notes for your reviewer**:

Requires https://github.com/kubernetes/kubernetes/pull/91381

/cc @pjbgf @hasheddan @evrardjp

**Does this PR introduce a user-facing change?**:

```release-note
- Added pod version skew strategy for seccomp profile to synchronize the deprecated annotations with the new API Server fields. Please see the corresponding section [in the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190717-seccomp-ga.md#version-skew-strategy) for more detailed explanations. 
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190717-seccomp-ga.md
```

### CODE REVIEW

File: pkg/apis/core/v1/conversion.go [hunk 1]
```diff
@@ -555,3 +563,109 @@ func Convert_core_PersistentVolumeSpec_To_v1_PersistentVolumeSpec(in *core.Persi
...
+				Type: v1.SeccompProfileUnconfined,
+			}
+		} else if annotation == v1.SeccompProfileRuntimeDefault {
+			out.Spec.SecurityContext.SeccompProfile = &v1.SeccompProfile{
+				Type: v1.SeccompProfileTypeDefault,
+			}
+		} else if strings.HasPrefix(annotation, localhostPrefix) {
+			out.Spec.SecurityContext.SeccompProfile = &v1.SeccompProfile{
+				Type:             v1.SeccompProfileLocalhost,
+				LocalhostProfile: &annotation,
```
Comment (pjbgf) [inline pkg/apis/core/v1/conversion.go:631]: I think the `annotation` would always be prefixed `localhost/` on this case. However, `LocalhostProfile` needs that removed, as the seccomp type now comes from type discriminator.

File: pkg/apis/core/v1/conversion.go [hunk 2]
```diff
@@ -555,3 +563,109 @@ func Convert_core_PersistentVolumeSpec_To_v1_PersistentVolumeSpec(in *core.Persi
...
+	if field != nil && !annotationAvailable {
+		switch field.Type {
+		case core.SeccompProfileUnconfined:
+			out.Annotations[v1.SeccompPodAnnotationKey] = profileUnconfined
+
+		case core.SeccompProfileTypeDefault:
+			out.Annotations[v1.SeccompPodAnnotationKey] = v1.SeccompProfileRuntimeDefault
+
+		case core.SeccompProfileLocalhost:
+			out.Annotations[v1.SeccompPodAnnotationKey] = localhostPrefix + *field.LocalhostProfile
```
Comment (pjbgf) [inline pkg/apis/core/v1/conversion.go:602]: I am not sure whether it is worth pointing out in comment (or doing the validation) that `field.LocalhostProfile` should not be nil at this point, as there is some validation that blocks that on creation.

Comment (saschagrunert) [reply pkg/apis/core/v1/conversion.go:602]: I'll add a check for that just to be sure. I think we should capture that.

File: pkg/apis/core/v1/conversion.go (hunk 1 above)
Comment (saschagrunert) [reply pkg/apis/core/v1/conversion.go:631]: Yes, changed to trim the prefix.

**Comment by saschagrunert** [comment]
Added the unit tests to bump the coverage of the new function to 100%.

File: pkg/apis/core/v1/conversion_test.go [hunk 3]
```diff
@@ -704,3 +705,271 @@ func Test_v1_NodeSpec_to_core_NodeSpec(t *testing.T) {
...
+			validation: func(t *testing.T, desc string, err error, pod *v1.Pod) {
+				require.NotNil(t, err, desc)
+			},
+		},
+		{
+			description: "Field type localhost and annotation does not match",
+			pod: &core.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{
+						v1.SeccompPodAnnotationKey: "not-matching",
```
Comment (pjbgf) [inline pkg/apis/core/v1/conversion_test.go:952]: May be worth to also have a case with annotation prefixed "localhost/" but with different profile name.

Comment (saschagrunert) [reply pkg/apis/core/v1/conversion_test.go:952]: Sure, why not. Added a test for that case.

**Comment by saschagrunert** [comment]
Seems unrelated:

```
STEP: Destroying namespace "pv-943" for this suite.
• Failure [346.655 seconds]
[sig-storage] PersistentVolumes
test/e2e/storage/utils/framework.go:23
  NFS
  test/e2e/storage/persistent_volumes.go:122
    with Single PV - PVC pairs
skipped 1722 lines unfold_more
[Fail] [sig-storage] PersistentVolumes NFS with Single PV - PVC pairs [It] create a PVC and a pre-bound PV: test write access 
test/e2e/storage/persistent_volumes.go:52
Ran 632 of 5098 Specs in 1046.741 seconds
FAIL! -- 631 Passed | 1 Failed | 0 Pending | 4466 Skipped
```

/test pull-kubernetes-e2e-kind

File: pkg/apis/core/v1/conversion.go [hunk 4]
```diff
@@ -555,3 +563,113 @@ func Convert_core_PersistentVolumeSpec_To_v1_PersistentVolumeSpec(in *core.Persi
 func Convert_v1_PersistentVolumeSpec_To_core_PersistentVolumeSpec(in *v1.PersistentVolumeSpec, out *core.PersistentVolumeSpec, s conversion.Scope) error {
 	return autoConvert_v1_PersistentVolumeSpec_To_core_PersistentVolumeSpec(in, out, s)
 }
+
+// HandleSeccompVersionSkew implements the version skew behavior described in:
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190717-seccomp-ga.md#version-skew-strategy
+func HandleSeccompVersionSkew(in *core.Pod, out *v1.Pod) error {
```
Comment (saschagrunert) [inline pkg/apis/core/v1/conversion.go:569]: @pjbgf what about the security context inside of the pods containers? Do we have to update/sync them as well?

Comment (pjbgf) [reply pkg/apis/core/v1/conversion.go:569]: That's correct, we need to extend the same logic for containers when the annotation `SeccompContainerAnnotationKeyPrefix` + container name is being used.

Comment (saschagrunert) [reply pkg/apis/core/v1/conversion.go:568]: Added the handling for the containers, too.

File: staging/src/k8s.io/kubectl/pkg/util/templates/markdown.go [hunk 5]
```diff
@@ -111,22 +111,24 @@ func (r *ASCIIRenderer) TableRow(out *bytes.Buffer, text []byte)
...
-func (r *ASCIIRenderer) Entity(out *bytes.Buffer, entity []byte)                          { r.fw(out, entity) }
-func (r *ASCIIRenderer) Smartypants(out *bytes.Buffer, text []byte)                       { r.fw(out, text) }
-func (r *ASCIIRenderer) DocumentHeader(out *bytes.Buffer)                                 {}
-func (r *ASCIIRenderer) DocumentFooter(out *bytes.Buffer)                                 {}
-func (r *ASCIIRenderer) TocHeaderWithAnchor(text []byte, level int, anchor string)        {}
-func (r *ASCIIRenderer) TocHeader(text []byte, level int)                                 {}
-func (r *ASCIIRenderer) TocFinalize()                                                     {}
+func (r *ASCIIRenderer) FootnoteItem(out *bytes.Buffer, name, text []byte, flags int) {
+	r.fw(out, text)
+}
```
Comment (hasheddan) [inline staging/src/k8s.io/kubectl/pkg/util/templates/markdown.go:116]: Are formatting changes like this expected?

Comment (saschagrunert) [reply staging/src/k8s.io/kubectl/pkg/util/templates/markdown.go:116]: I'm not sure, I updated them via `make update`. Let me revert those changes and try it again.

**Comment by saschagrunert** [comment]
Hm... :thinking: 
/test pull-kubernetes-e2e-gce

File: staging/src/k8s.io/kubectl/pkg/util/templates/markdown.go (hunk 5 above)
Comment (hasheddan) [reply staging/src/k8s.io/kubectl/pkg/util/templates/markdown.go:116]: I'm not sure how much it really matters, but it does appear to be adding some noise that isn't related to this change

Comment (saschagrunert) [reply staging/src/k8s.io/kubectl/pkg/util/templates/markdown.go]: Should be fine now. :)

**Comment by saschagrunert** [comment]
Have to look into the test failures in the next step.

**Comment by saschagrunert** [comment]
The base PR has changed. I'll rebase on top if there is an agreement about the architecture there.

/wip

File: pkg/apis/core/v1/conversion.go [hunk 6]
```diff
@@ -539,3 +546,146 @@ func Convert_core_PersistentVolumeSpec_To_v1_PersistentVolumeSpec(in *core.Persi
...
+		// Ensure that the output security context is not nil and set the new
+		// profile
+		if out.Spec.SecurityContext == nil {
+			out.Spec.SecurityContext = &v1.PodSecurityContext{}
+		}
+		out.Spec.SecurityContext.SeccompProfile = profile
+	}
+
+	// Handle the containers of the pod
+	for i, ctr := range in.Spec.Containers {
```
Comment (saschagrunert) [inline pkg/apis/core/v1/conversion.go:577]: Note: Use `podutil.VisitContainers` like in https://github.com/kubernetes/kubernetes/pull/91381/files#diff-81c0c668d90c62bb5b7a2334c09c7e50R3781

**Comment by saschagrunert** [comment]
Just rebased on top of #91381 without any changes in my commit. I'm planning to adapt the logic this week.

File: pkg/apis/core/v1/conversion.go [hunk 7]
```diff
@@ -382,6 +385,10 @@ func Convert_core_Pod_To_v1_Pod(in *core.Pod, out *v1.Pod, s conversion.Scope) e
 	// remove this once the oldest supported kubelet no longer honors the annotations over the field.
 	out.Annotations = dropInitContainerAnnotations(out.Annotations)
 
+	if err := HandleSeccompVersionSkew(in, out); err != nil {
+		return errors.Wrap(err, "handling seccomp version skew")
+	}
```
Comment (liggitt) [inline pkg/apis/core/v1/conversion.go:390]: rather than doing this in conversion, I expected it to be contained in pkg/registry/core/pod/strategy.go and called from `podStrategy#PrepareForCreate`

File: pkg/apis/core/v1/conversion.go (hunk 6 above)
Comment (liggitt) [inline pkg/apis/core/v1/conversion.go:577]: use VisitContainers to make sure you iterate over all container types (containers, initContainers, ephemeralContainers)

File: pkg/apis/core/v1/conversion.go (hunk 7 above)
Comment (saschagrunert) [reply pkg/apis/core/v1/conversion.go:390]: That’s a great hint and saves me time, thank you!

File: pkg/apis/core/v1/conversion.go [hunk 8]
```diff
@@ -539,3 +546,146 @@ func Convert_core_PersistentVolumeSpec_To_v1_PersistentVolumeSpec(in *core.Persi
 func Convert_v1_PersistentVolumeSpec_To_core_PersistentVolumeSpec(in *v1.PersistentVolumeSpec, out *core.PersistentVolumeSpec, s conversion.Scope) error {
 	return autoConvert_v1_PersistentVolumeSpec_To_core_PersistentVolumeSpec(in, out, s)
 }
+
+// HandleSeccompVersionSkew implements the version skew behavior described in:
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190717-seccomp-ga.md#version-skew-strategy
+func HandleSeccompVersionSkew(in *core.Pod, out *v1.Pod) error {
+	// Handle the pod
```
Comment (liggitt) [inline pkg/apis/core/v1/conversion.go:553]: we only care about the case where they only provided one of the annotation and the field, right? would this be easier to follow like this:

```go
annotation, hasAnnotation := pod.Annotations[v1.SeccompPodAnnotationKey]

field, hasField := (*core.SeccompProfile)(nil), false
if pod.Spec.SecurityContext != nil && pod.Spec.SecurityContext.SeccompProfile != nil {
  field = pod.Spec.SecurityContext.SeccompProfile
  hasField = true
}

if hasAnnotation != hasField {
  // initialize parents if necessary
  if pod.Annotations == nil {
    pod.Annotations = map[string]string{}
  }
  if pod.Spec.SecurityContext == nil {
    pod.Spec.SecurityContext = &v1.PodSecurityContext{}
  }
  // assign computed values
  pod.Annotations[v1.SeccompPodAnnotationKey], pod.Spec.SecurityContext.SeccompProfile = syncSeccompAnnotationAndField(annotation, field)
}
```

(and then something similar per-container inside VisitContainers)

File: pkg/apis/core/v1/conversion.go [hunk 9]
```diff
@@ -539,3 +546,146 @@ func Convert_core_PersistentVolumeSpec_To_v1_PersistentVolumeSpec(in *core.Persi
...
+				out.Spec.Containers[i].SecurityContext = &v1.SecurityContext{}
+			}
+			out.Spec.Containers[i].SecurityContext.SeccompProfile = profile
+		}
+	}
+
+	return nil
+}
+
+func syncSeccompAnnotationAndFieldForPod(annotationKey, annotationValue string, field *core.SeccompProfile, pod *v1.Pod) (*v1.SeccompProfile, error) {
```
Comment (liggitt) [inline pkg/apis/core/v1/conversion.go:613]: I think making this a pure function would make it easier to review and test... as it is, returning one of the computed values and storing the other by mutating the passed-in pod's annotations is a bit hard to follow

File: pkg/apis/core/v1/conversion.go [hunk 10]
```diff
@@ -539,3 +546,146 @@ func Convert_core_PersistentVolumeSpec_To_v1_PersistentVolumeSpec(in *core.Persi
...
+			localhostProfile := strings.TrimPrefix(annotationValue, v1.SeccompLocalhostProfileNamePrefix)
+			profile.Type = v1.SeccompProfileLocalhost
+			profile.LocalhostProfile = &localhostProfile
+		}
+		return profile, nil
+	}
+
+	// If both seccomp annotations and fields are specified, the values must
+	// match.
+	if field != nil && annotationValue != "" {
```
Comment (liggitt) [inline pkg/apis/core/v1/conversion.go:662]: this is enforced in validation, don't need to enforce this here

File: pkg/apis/core/v1/conversion.go [hunk 11]
```diff
@@ -539,3 +546,146 @@ func Convert_core_PersistentVolumeSpec_To_v1_PersistentVolumeSpec(in *core.Persi
...
+	// If only seccomp annotations are specified, copy the values into the
+	// corresponding fields. This ensures that existing applications continue
+	// to enforce seccomp, and prevents the kubelet from needing to resolve
+	// annotations & fields.
+	if field == nil && annotationValue != "" {
+		profile := &v1.SeccompProfile{}
+		if annotationValue == v1.SeccompProfileNameUnconfined {
+			profile.Type = v1.SeccompProfileUnconfined
+
+		} else if annotationValue == v1.SeccompProfileRuntimeDefault {
```
Comment (liggitt) [inline pkg/apis/core/v1/conversion.go:649]: should this also honor DeprecatedSeccompProfileDockerDefault?

Comment (saschagrunert) [reply pkg/apis/core/v1/conversion.go]: Yes, added it as well.

File: pkg/apis/core/v1/conversion.go (hunk 10 above)
Comment (saschagrunert) [reply pkg/apis/core/v1/conversion.go]: Okay, removed the whole block.

File: pkg/apis/core/v1/conversion.go (hunk 9 above)
Comment (saschagrunert) [reply pkg/apis/core/v1/conversion.go]: Sounds good, I'll change as suggested. :+1:

File: pkg/apis/core/v1/conversion.go (hunk 8 above)
Comment (saschagrunert) [reply pkg/apis/core/v1/conversion.go]: Yes, thank you for the hint. Changed as suggested.

File: pkg/apis/core/v1/conversion.go (hunk 6 above)
Comment (saschagrunert) [reply pkg/apis/core/v1/conversion.go]: Okay, switched to the VisitContainers utility function.

**Comment by saschagrunert** [comment]
The API server segfaults because of the field validation. The error list returned after the seccomp field modification contains a single nil error `[<nil>]` which might needs to be fixed in the original PR.

**Comment by liggitt** [comment]
> The API server segfaults because of the field validation

Out of curiosity, is that actually killing the apiserver process, or just affecting the individual request? panics are intended to be caught at the http handler level and limited to affecting the individual request.

**Comment by saschagrunert** [comment]
> > The API server segfaults because of the field validation
> 
> Out of curiosity, is that actually killing the apiserver process, or just affecting the individual request? panics are intended to be caught at the http handler level and limited to affecting the individual request.

It's just reporting an 500 internal server error and the API server still continues to live, so I think the error handling path is fine. It was a bit hard to debug because I had no direct link between the `nil` error unwrapping and where the actual error occurs. But I think this was just a luxury problem…

File: pkg/registry/core/pod/strategy.go [hunk 12]
```diff
@@ -569,3 +572,120 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
...
+		field = pod.Spec.SecurityContext.SeccompProfile
+		hasField = true
+	}
+
+	if hasAnnotation != hasField {
+		// assign computed values
+		newAnnotation, newField := syncSeccompAnnotationAndField(annotation, field)
+		if newAnnotation != "" {
+			pod.Annotations[v1.SeccompPodAnnotationKey] = newAnnotation
+		}
```
Comment (liggitt) [inline pkg/registry/core/pod/strategy.go:596]:
```suggestion
		if newAnnotation != "" {
			if pod.Annotations == nil {
				pod.Annotations = map[string]string{}
			}
			pod.Annotations[v1.SeccompPodAnnotationKey] = newAnnotation
		}
```

File: pkg/registry/core/pod/strategy.go [hunk 13]
```diff
@@ -569,3 +572,120 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
...
+				annotation    string
+				hasAnnotation bool
+			)
+			for key, value := range pod.Annotations {
+				if key == api.SeccompContainerAnnotationKeyPrefix+ctr.Name {
+					annotation = value
+					hasAnnotation = true
+					break
+				}
+			}
```
Comment (liggitt) [inline pkg/registry/core/pod/strategy.go:619]:
```suggestion
			key := api.SeccompContainerAnnotationKeyPrefix+ctr.Name
			annotation, hasAnnotation := pod.Annotations[key]
```

File: pkg/registry/core/pod/strategy.go [hunk 14]
```diff
@@ -569,3 +572,120 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
...
+	if field != nil && annotation == "" {
+		switch field.Type {
+		case api.SeccompProfileTypeUnconfined:
+			return v1.SeccompProfileNameUnconfined, nil
+
+		case api.SeccompProfileTypeRuntimeDefault:
+			return v1.SeccompProfileRuntimeDefault, nil
+
+		case api.SeccompProfileTypeLocalhost:
+			return v1.SeccompLocalhostProfileNamePrefix + *field.LocalhostProfile, nil
```
Comment (liggitt) [inline pkg/registry/core/pod/strategy.go:664]: this happens pre-validation, so field.LocalhostProfile can be nil... don't dereference without checking

File: pkg/registry/core/pod/strategy.go [hunk 15]
```diff
@@ -569,3 +572,120 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
...
+// field and returns the synchronized result: The annotation will be converted
+// in an appropriate profile and vice versa.
+func syncSeccompAnnotationAndField(annotation string, field *api.SeccompProfile) (string, *api.SeccompProfile) {
+	// If only seccomp fields are specified, add the corresponding annotations.
+	// This ensures that the fields are enforced even if the node version
+	// trails the API version
+	if field != nil && annotation == "" {
+		switch field.Type {
+		case api.SeccompProfileTypeUnconfined:
+			return v1.SeccompProfileNameUnconfined, nil
```
Comment (liggitt) [inline pkg/registry/core/pod/strategy.go:658]: I had envisioned returning both the annotation value and the field value in all cases. Doing it this way is fine, but if this is going to be implemented in a way that only returns one of the values, it would probably make more sense as two separate methods:

```go
func seccompAnnotationForField(field *api.SeccompProfile) string {
... assume field is non-nil ...
}

func seccompFieldForAnnotation(annotation string) *api.SeccompProfile {
... assume annotation is non-empty ...
}
```

and be called like this:
```go
switch {
  case hasAnnotation && !hasField:
    if pod.Spec.SecurityContext == nil {
      pod.Spec.SecurityContext = &api.PodSecurityContext{}
    }
    pod.Spec.SecurityContext.SeccompProfile = seccompFieldForAnnotation(annotation)

  case hasField && !hasAnnotation:
    if pod.Annotations == nil {
      pod.Annotations = map[string]string{}
    }
    pod.Annotations[key] = seccompAnnotationForField(field)
}
```

File: pkg/registry/core/pod/strategy.go [hunk 16]
```diff
@@ -569,3 +572,120 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
...
+			return "", &api.SeccompProfile{Type: api.SeccompProfileTypeUnconfined}
+
+		} else if annotation == api.SeccompProfileRuntimeDefault || annotation == api.DeprecatedSeccompProfileDockerDefault {
+			return "", &api.SeccompProfile{Type: api.SeccompProfileTypeRuntimeDefault}
+
+		} else if strings.HasPrefix(annotation, v1.SeccompLocalhostProfileNamePrefix) {
+			localhostProfile := strings.TrimPrefix(annotation, v1.SeccompLocalhostProfileNamePrefix)
+			return "", &api.SeccompProfile{
+				Type:             api.SeccompProfileTypeLocalhost,
+				LocalhostProfile: &localhostProfile,
```
Comment (liggitt) [inline pkg/registry/core/pod/strategy.go:683]: this is called pre-validation ... should we handle a zero-length localhostProfile by omitting this?

**Comment by liggitt** [comment]
> The error list returned after the seccomp field modification contains a single nil error `[<nil>]` which might needs to be fixed in the original PR (or in this PR).

commented on the two places in the original PR that need to fix t his

File: pkg/registry/core/pod/strategy.go (hunk 16 above)
Comment (saschagrunert) [reply pkg/registry/core/pod/strategy.go:689]: Okay, added a check and unit test for `localhostProfile != ""`.

File: pkg/registry/core/pod/strategy.go (hunk 15 above)
Comment (saschagrunert) [reply pkg/registry/core/pod/strategy.go]: I split-up the function into two. They can still return `nil` or `""` if the localhost profile is invalid, so I have to check for the result again.

File: pkg/registry/core/pod/strategy.go (hunk 14 above)
Comment (saschagrunert) [reply pkg/registry/core/pod/strategy.go]: Added a check, in that case we now return `""` for the annotation (which is treated as invalid).

File: pkg/registry/core/pod/strategy.go (hunk 13 above)
Comment (saschagrunert) [reply pkg/registry/core/pod/strategy.go]: Much easier, changed as suggested. :+1:

File: pkg/registry/core/pod/strategy.go (hunk 12 above)
Comment (saschagrunert) [reply pkg/registry/core/pod/strategy.go]: Yes, changed as suggested.

File: pkg/registry/core/pod/strategy.go [hunk 17]
```diff
@@ -569,3 +572,120 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
...
+				field = ctr.SecurityContext.SeccompProfile
+				hasField = true
+			}
+
+			if hasAnnotation != hasField {
+				// initialize parents if necessary
+				// assign computed values
+				newAnnotation, newField := syncSeccompAnnotationAndField(annotation, field)
+				if newAnnotation != "" {
+					pod.Annotations[v1.SeccompContainerAnnotationKeyPrefix+ctr.Name] = newAnnotation
```
[the remark this answers is not in the document]
Comment (saschagrunert) [reply pkg/registry/core/pod/strategy.go]: Should be fixed now.

File: pkg/registry/core/pod/strategy.go [hunk 18]
```diff
@@ -569,3 +572,120 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
...
 	return container, nil
 }
+
+// applySeccompVersionSkew implements the version skew behavior described in:
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190717-seccomp-ga.md#version-skew-strategy
+func applySeccompVersionSkew(pod *api.Pod) {
+	// initialize parents if necessary
+	if pod.Annotations == nil {
+		pod.Annotations = map[string]string{}
+	}
```
[the remark this answers is not in the document]
Comment (saschagrunert) [reply pkg/registry/core/pod/strategy.go]: Changed all code paths to only init if necessary.

**Comment by saschagrunert** [comment]
Ready for review again, please take another look @liggitt

File: pkg/registry/core/pod/strategy.go [hunk 19]
```diff
@@ -568,3 +572,126 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
...
+
+	// Handle the containers of the pod
+	podutil.VisitContainers(&pod.Spec, podutil.AllFeatureEnabledContainers(),
+		func(ctr *api.Container, _ podutil.ContainerType) bool {
+			// get possible annotation and field
+			key := api.SeccompContainerAnnotationKeyPrefix + ctr.Name
+			annotation, hasAnnotation := pod.Annotations[key]
+
+			field := (*api.SeccompProfile)(nil)
```
Comment (liggitt) [inline pkg/registry/core/pod/strategy.go:616]: reset `hasField` to false here or a `true` value from a previous iteration or from the pod-level check could be inherited

Comment (liggitt) [reply pkg/registry/core/pod/strategy.go:616]: add a unit test that would have caught this (pod-level seccompProfile field, container-level seccomp annotation)

File: pkg/registry/core/pod/strategy.go [hunk 20]
```diff
@@ -568,3 +572,126 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
...
+		return v1.SeccompProfileRuntimeDefault
+
+	case api.SeccompProfileTypeLocalhost:
+		if field.LocalhostProfile != nil {
+			return v1.SeccompLocalhostProfileNamePrefix + *field.LocalhostProfile
+		}
+	}
+
+	// we can only reach this code path if the LocalhostProfile is nil but the
+	// provided field type is SeccompProfileTypeLocalhost
```
Comment (liggitt) [inline pkg/registry/core/pod/strategy.go:667]: or if an unrecognized type is specified

File: pkg/registry/core/pod/strategy.go [hunk 21]
```diff
@@ -568,3 +572,126 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
...
+// seccomp profile field.
+func seccompFieldForAnnotation(annotation string) *api.SeccompProfile {
+	// If only seccomp annotations are specified, copy the values into the
+	// corresponding fields. This ensures that existing applications continue
+	// to enforce seccomp, and prevents the kubelet from needing to resolve
+	// annotations & fields.
+	if annotation == v1.SeccompProfileNameUnconfined {
+		return &api.SeccompProfile{Type: api.SeccompProfileTypeUnconfined}
+
+	} else if annotation == api.SeccompProfileRuntimeDefault || annotation == api.DeprecatedSeccompProfileDockerDefault {
```
Comment (liggitt) [inline pkg/registry/core/pod/strategy.go:681]: nit: no need to `else if` in return cases:

```go
if annotation == v1.SeccompProfileNameUnconfined {
	return &api.SeccompProfile{Type: api.SeccompProfileTypeUnconfined}
}

if annotation == api.SeccompProfileRuntimeDefault || annotation == api.DeprecatedSeccompProfileDockerDefault {
	return &api.SeccompProfile{Type: api.SeccompProfileTypeRuntimeDefault}
}

if strings.HasPrefix(annotation, v1.SeccompLocalhostProfileNamePrefix) {
	...
}

return nil
```

File: pkg/registry/core/pod/strategy.go [hunk 22]
```diff
@@ -568,3 +572,126 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
...
+		if localhostProfile != "" {
+			return &api.SeccompProfile{
+				Type:             api.SeccompProfileTypeLocalhost,
+				LocalhostProfile: &localhostProfile,
+			}
+		}
+	}
+
+	// we can only reach this code path if the localhostProfile name has a zero
+	// length
```
Comment (liggitt) [inline pkg/registry/core/pod/strategy.go:695]: or if the annotation has an unrecognized value

**Comment by liggitt** [comment]
Looks like the nil error issue was fixed in the original PR. This is very close, noted one bug and a few nits.

File: pkg/registry/core/pod/strategy.go (hunk 19 above)
Comment (saschagrunert) [reply pkg/registry/core/pod/strategy.go]: Fixed, added two unit tests, one for field -> annotation  and another one vice versa.

File: pkg/registry/core/pod/strategy.go (hunk 20 above)
Comment (saschagrunert) [reply pkg/registry/core/pod/strategy.go]: Added the comment. :+1:

File: pkg/registry/core/pod/strategy.go (hunk 21 above)
Comment (saschagrunert) [reply pkg/registry/core/pod/strategy.go]: Fixed as suggested.

File: pkg/registry/core/pod/strategy.go (hunk 22 above)
Comment (saschagrunert) [reply pkg/registry/core/pod/strategy.go]: Added the comment. :+1:

**Comment by saschagrunert** [comment]
> Looks like the nil error issue was fixed in the original PR. This is very close, noted one bug and a few nits.

Addressed the open issues, I think I have to rebase on top of the master if https://github.com/kubernetes/kubernetes/pull/91381 got merged.

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/91381 is merged, can rebase on master now

**Comment by saschagrunert** [comment]
> #91381 is merged, can rebase on master now

Done :heavy_check_mark:

**Comment by saschagrunert** [comment]
If this gets merged then I will continue to add e2e tests for the API migration in https://github.com/kubernetes/kubernetes/pull/91442
